File: java/ValidParentheses.java
// 20. Valid Parentheses
// Accepted 8ms

class Solution {
    public boolean isValid(String s) {
        List<Character> stk = new ArrayList<Character>();
        char c;
        
        for(int k = 0; k < s.length(); k++) {
            c = s.charAt(k);
            
            if(c == '(' || c == '[' || c == '{')
                stk.add(c);
            else {
                if(c == ')' && (stk.size() == 0 || stk.get(stk.size() - 1) != '('))
                    return false;
                if(c == ']' && (stk.size() == 0 || stk.get(stk.size() - 1) != '['))
                    return false;
                if(c == '}' && (stk.size() == 0 || stk.get(stk.size() - 1) != '{'))
                    return false;
                
                stk.remove(stk.size() - 1);
            }
        }
        
        return (stk.size() == 0);
    }
}

File: src/problems/020. Valid Parentheses/ValidParentheses.py
# 20. Valid Parentheses
# Accepted 32ms

class Solution:
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        stack = []
        for c in s:
            if c == '(' or c == '{' or c == '[':
                stack.append(c)
            else:
                if len(stack) > 0:
                    top = stack.pop()
                    if not ((c == ')' and top == '(') or (c == '}' and top == '{') or (c == ']' and top == '[')):
                        return False
                else:
                    return False
        
        if len(stack) > 0:
            return False
        return True
